dll/cpl/input/settings_page.c

```diff
@@ -10,195 +10,358 @@
 #include "layout_list.h"
 #include "locale_list.h"
 #include "input_list.h"
+#include <shellapi.h>
+
+static INT s_nAliveLeafCount = 0;
+static INT s_nRootCount = 0;
+static INT s_iKeyboardImage = -1;
+static INT s_iDotImage = -1;
 
 static HICON
-CreateLayoutIcon(LPWSTR szLayout, BOOL bIsDefault)
+CreateLayoutIcon(LANGID LangID)
 {
-    INT width = GetSystemMetrics(SM_CXSMICON) * 2;
-    INT height = GetSystemMetrics(SM_CYSMICON);
+    WCHAR szBuf[4];
     HDC hdc;
-    HDC hdcsrc;
-    HBITMAP hBitmap;
-    HICON hIcon = NULL;
-
-    hdcsrc = GetDC(NULL);
-    hdc = CreateCompatibleDC(hdcsrc);
-    hBitmap = CreateCompatibleBitmap(hdcsrc, width, height);
-
-    ReleaseDC(NULL, hdcsrc);
-
-    if (hdc && hBitmap)
+    HBITMAP hbmColor, hbmMono, hBmpOld;
+    RECT rect;
+    HFONT hFontOld, hFont;
+    ICONINFO IconInfo;
+    HICON hIcon;
+    LOGFONTW lf;
+    BITMAPINFO bmi;
+    INT cxIcon = GetSystemMetrics(SM_CXSMICON);
+    INT cyIcon = GetSystemMetrics(SM_CYSMICON);
+
+    /* Getting "EN", "FR", etc. from English, French, ... */
+    if (!GetLocaleInfoW(LangID, LOCALE_SABBREVLANGNAME | LOCALE_NOUSEROVERRIDE,
+                        szBuf, ARRAYSIZE(szBuf)))
     {
-        HBITMAP hBmpNew;
-
-        hBmpNew = CreateBitmap(width, height, 1, 1, NULL);
-        if (hBmpNew)
-        {
-            LOGFONTW lf;
-
-            if (SystemParametersInfoW(SPI_GETICONTITLELOGFONT, sizeof(lf), &lf, 0))
-            {
-                ICONINFO IconInfo;
-                HFONT hFont;
-
-                hFont = CreateFontIndirectW(&lf);
-
-                if (hFont != NULL)
-                {
-                    HBITMAP hBmpOld;
+        StringCchCopyW(szBuf, ARRAYSIZE(szBuf), L"??");
+    }
+    szBuf[2] = UNICODE_NULL; /* Truncate the identifiers to two characters: "ENG" --> "EN" etc. */
+
+    /* Prepare for DIB (device-independent bitmap) */
+    ZeroMemory(&bmi, sizeof(bmi));
+    bmi.bmiHeader.biSize = sizeof(bmi.bmiHeader);
+    bmi.bmiHeader.biWidth = cxIcon;
+    bmi.bmiHeader.biHeight = cyIcon;
+    bmi.bmiHeader.biPlanes = 1;
+    bmi.bmiHeader.biBitCount = 24;
```

Comment:
I wonder whether this should always be "24", or be adjusted depending on the underlying DC (that is however currently created below), with `GetDeviceCaps(hdc, BITSPIXEL);`

Reply:
The 24-bpp DIB (device-independent bitmap) is a full color bitmap. So it doesn't need the color table if conversion is correct. The DDB (device-dependent bitmap) has a fear that conversion might be failed if the device was changed.

Reply:
See also: https://github.com/reactos/reactos/pull/4723#discussion_r981796819